File: apps/ios/ios-rust-test/UI/Card/CardViewModel.swift
import Foundation

/// Filtros de **texto**, no validaciones: quien decide si el número pasa Luhn o si el hex
/// es descifrable es el core.
private let digitsOnly = /^[0-9]*$/
private let lowercaseHex = /^[0-9a-f]*$/

@MainActor
@Observable
final class CardViewModel {
    private(set) var state = CardUiState()

    private let core: CoreFinanciero
    private let messages: ContractMessages
    private let keyHex: String
    private let nonceHex: String

    init(core: CoreFinanciero, contract: ContractSource, messages: ContractMessages) {
        self.core = core
        self.messages = messages
        // El nonce es FIJO a propósito: es lo que hace que las cuatro plataformas produzcan
        // el mismo hex y se pueda comparar en la demo. En producción sería catastrófico.
        keyHex = contract.demoKeyHex()
        nonceHex = contract.demoNonceHex()
    }

    // MARK: - Entrada del usuario

    /// Escribe el estado **siempre**, aunque el valor no cambie: un `set` que no escribe deja la
    /// tecla rechazada a la vista. La explicación completa está en `TransferViewModel.amountChanged`.
    func numberChanged(_ value: String) {
        let accepted = value.wholeMatch(of: digitsOnly) != nil
        state.number = accepted ? value : state.number
        if accepted { state.encryptError = nil }
    }

    /// Ídem `numberChanged`.
    func pastedHexChanged(_ value: String) {
        let accepted = value.wholeMatch(of: lowercaseHex) != nil
        state.pastedHex = accepted ? value : state.pastedHex
        if accepted { state.decryptError = nil }
    }

    // MARK: - Acciones

    func validateAndEncrypt() {
        do {
            let card = try core.validateCard(number: state.number)
            let hex = try core.encrypt(text: state.number, keyHex: keyHex, nonceHex: nonceHex)
            let back = try core.decrypt(ciphertextHex: hex, keyHex: keyHex, nonceHex: nonceHex)
            state.brand = card.brand
            state.masked = card.masked
            state.cipherHex = hex
            state.decrypted = back
            state.encryptError = nil
        } catch let e as DomainError {
            clearEncryptResult()
            state.encryptError = messages.userMessage(e)
        } catch {
            clearEncryptResult()
            state.encryptError = messages.userMessage(error)
        }
    }

    func decryptPasted() {
        do {
            state.recovered = try core.decrypt(
                ciphertextHex: state.pastedHex,
                keyHex: keyHex,
                nonceHex: nonceHex
            )
            state.decryptError = nil
        } catch let e as DomainError {
            state.recovered = ""
            state.decryptError = messages.userMessage(e)
        } catch {
            state.recovered = ""
            state.decryptError = messages.userMessage(error)
        }
    }

    private func clearEncryptResult() {
        state.brand = ""
        state.masked = ""
        state.cipherHex = ""
        state.decrypted = ""
    }
}

File: apps/android/app/src/main/java/dev/tohure/android_rust_test/ui/card/CardViewModel.kt
package dev.tohure.android_rust_test.ui.card

import androidx.lifecycle.ViewModel
import dev.tohure.android_rust_test.adapter.ContractMessages
import dev.tohure.android_rust_test.adapter.CoreFinanciero
import dev.tohure.android_rust_test.contract.ContractSource
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import uniffi.core_financiero.DomainException

class CardViewModel(
    private val core: CoreFinanciero,
    private val contract: ContractSource,
    private val messages: ContractMessages,
) : ViewModel() {
    private val _uiState = MutableStateFlow(CardUiState())
    val uiState: StateFlow<CardUiState> = _uiState.asStateFlow()

    // ── Entrada del usuario ───────────────────────────────────────────────────

    fun numberChanged(value: String) {
        // Solo dígitos: filtro de texto. Luhn lo valida el core, no esta app.
        if (!value.all(Char::isDigit)) return
        _uiState.value = _uiState.value.copy(number = value)
        clearError()
    }

    // ── Acciones ──────────────────────────────────────────────────────────────

    fun validateAndEncrypt() {
        val number = _uiState.value.number
        core.validateCard(number)
            .mapCatching { card ->
                val hex = core.encrypt(number, contract.demoKeyHex(), contract.demoNonceHex())
                    .getOrThrow()
                card to hex
            }.onSuccess { (card, hex) ->
                _uiState.value = _uiState.value.copy(
                    brand = card.brand,
                    masked = card.masked,
                    cipherHex = hex,
                    error = null,
                )
            }.onFailure { e ->
                _uiState.value = _uiState.value.copy(
                    brand = "",
                    masked = "",
                    cipherHex = "",
                    error = (e as? DomainException)?.let(messages::userMessage) ?: e.toString(),
                )
            }
    }

    fun clearError() { _uiState.value = _uiState.value.copy(error = null) }
}
